src/restic/backend/b2/b2.go

```diff
@@ -0,0 +1,288 @@
+	objName := be.b2path(h.Type, h.Name)
+
+	obj := be.bucket.Object(objName)
+	if err != nil {
+		debug.Log("  err %v", err)
+		return 0, errors.Wrap(err, "bucket.Object")
+	}
+
+	info, err := obj.Attrs(be.context)
+	if err != nil {
+		return 0, errors.Wrap(err, "obj.Attrs")
+	}
+
+	// handle negative offsets
+	if off < 0 {
+		// if the negative offset is larger than the object itself, read from
+		// the beginning.
+		if -off > info.Size {
+			off = 0
+		} else {
+			// otherwise compute the offset from the end of the file.
+			off = info.Size + off
+		}
+	}
+
+	// return an error if the offset is beyond the end of the file
+	if off > info.Size {
+		return 0, errors.Wrap(io.EOF, "")
+	}
+
+	var nextError error
+
+	// manually create an io.ErrUnexpectedEOF
+	if off+int64(len(p)) > info.Size {
+		newlen := info.Size - off
+		p = p[:newlen]
+
+		nextError = io.ErrUnexpectedEOF
+
+		debug.Log("  capped buffer to %v byte", len(p))
+	}
+
+	r := obj.NewRangeReader(be.context, off, int64(len(p)))
+	defer r.Close()
+
+	bufsize := int64(len(p))
+	var pos int64 = 0
+
+	// loop to read chunks until the buffer is full or EOF is reached
+	for {
+		n, err = r.Read(p[pos:])
+		pos += int64(n)
+		if pos == info.Size-off || pos == bufsize || err != nil {
+			if errors.Cause(err) == io.EOF {
+				err = nil
+			}
+			break
+		}
+	}
+
+	if err == nil {
+		err = nextError
+	}
+
+	return int(pos), err
+}
+
+// Save stores data in the backend at the handle.
+func (be b2) Save(h restic.Handle, p []byte) (err error) {
+	debug.Log("Save: %v with len %d", h, len(p))
+	if err := h.Valid(); err != nil {
+		return err
+	}
+
+	objName := be.b2path(h.Type, h.Name)
+
+	obj := be.bucket.Object(objName)
+	if err != nil {
+		debug.Log("  err %v", err)
+		return errors.Wrap(err, "bucket.Object")
+	}
+
+	_, err = obj.Attrs(be.context)
+	if err == nil {
+		debug.Log("  %v already exists", h)
+		return errors.New("key already exists")
+	}
+
+	debug.Log("  Writing to %s", objName)
+	w := obj.NewWriter(be.context)
+	n, err := w.Write(p)
+
+	if err != nil {
+		debug.Log("  Write error %v -> %v bytes, err %#v", objName, n, err)
+		w.Close()
+		return err
+	}
+
+	debug.Log("  Write successful %v -> %v bytes", objName, n)
+	w.Close()
+	return nil
+}
+
+// Stat returns information about a blob.
+func (be b2) Stat(h restic.Handle) (bi restic.FileInfo, err error) {
+	debug.Log("Stat: %v", h)
+	objName := be.b2path(h.Type, h.Name)
+	obj := be.bucket.Object(objName)
+	info, err := obj.Attrs(be.context)
+	if err != nil {
+		debug.Log("Attrs() err %v", err)
+		return restic.FileInfo{}, errors.Wrap(err, "Stat")
+	}
+	return restic.FileInfo{Size: info.Size}, nil
+}
+
+// Test returns true if a blob of the given type and name exists in the backend.
+func (be *b2) Test(t restic.FileType, name string) (bool, error) {
+	found := false
+	objName := be.b2path(t, name)
+	obj := be.bucket.Object(objName)
+	_, err := obj.Attrs(be.context)
+	if err == nil {
```

Comment:
IIRC This might result in a problem if the file is currently being uploaded.

This should additionally check if `attrs.Status == b2.Uploaded`, otherwise it might return true for files that haven't been fully uploaded yet.